3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/adapters/registry/RemoteAdapter.scala

```diff
@@ -0,0 +1,103 @@
+package com.snowplowanalytics
+package snowplow
+package enrich
+package common
+package adapters
+package registry
+
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream, ObjectInputStream, ObjectOutputStream}
+import java.util.Base64
+
+import com.snowplowanalytics.iglu.client.Resolver
+import com.snowplowanalytics.snowplow.enrich.common.loaders.CollectorPayload
+import com.softwaremill.sttp._
+import org.slf4j.LoggerFactory
+import scalaz.Scalaz._
+
+import scala.concurrent.duration.FiniteDuration
+
+/**
+ * An adapter for an enrichment that is handled by a remote webservice.
+ *
+ * @constructor create a new client to talk to the given remote webservice.
+ * @param remoteUrl the url of the remote webservice, e.g. http://localhost/myEnrichment
+ * @param timeout max duration of each connection attempt, and of the wait for each response from the remote
+ */
+class RemoteAdapter(val remoteUrl: String, val timeout: FiniteDuration) extends Adapter {
+
+  private lazy val log = LoggerFactory.getLogger(getClass)
+
+  val remoteUri = uri"$remoteUrl"
+  implicit val backend = HttpURLConnectionBackend(
+    options = SttpBackendOptions.connectionTimeout(timeout)
+  )
+
+  /**
+   * POST the given payload to the remote webservice,
+   * wait for it to respond with an Either[List[String], List[RawEvent] ],
+   * and return that as a ValidatedRawEvents
+   *
+   * @param payload The CollectorPaylod containing one or more
+   *        raw events as collected by a Snowplow collector
+   * @param resolver (implicit) The Iglu resolver used for
+   *        schema lookup and validation
+   * @return a Validation boxing either a NEL of RawEvents on
+   *         Success, or a NEL of Failure Strings
+   */
+  def toRawEvents(payload: CollectorPayload)(implicit resolver: Resolver): ValidatedRawEvents =
+    httpPost(remoteUrl, RemoteAdapter.serializeToBase64(payload)) match {
+      case Left(errmsg) =>
+        errmsg.failNel
+      case Right(bodyAsString) =>
+        RemoteAdapter
+          .deserializeFromBase64(bodyAsString)
+          .asInstanceOf[Either[List[String], List[RawEvent]]] match {
```

Comment:
I'm not too sure about the input/output of the adapter, could we find a format better-suited for http?

Reply:
What is your suggestion? The choice is based on the fact that input is an unstructured text and output should be in format of `RawEvent` to be compatible with the next steps.

Reply:
In my mind the adapter could take care of transforming the input before forwarding the collector payload and then turn the result into a RawEvent when receiving it

Reply:
Got it. I think I should pull out some of the remote-adapter-example functionalities and put them here. So basically the steps would be these:
1) `RemoteAdapter` extracts the collector-specific information in the payload, and just sends the business related values/params to the webservice.
2) `Webservice` does the work to transfer the data to well-formed Snowplow event.
3) `RemoteAdapter` recives the payload and merges it with step1 data into a `RawEvent` payload.

Reply:
looks good on posting the collector payload, however why receive base64 and not directly json?

Reply:
encoding and binary data. We can try json and see what would happen. so shall I remove base64?

Reply:
yup I'd rather we go json out json in, created the repo: https://github.com/snowplow-incubator/remote-adapter-example/